pkg/cmd/roachtest/tests/mixed_version_multi_region.go

```diff
@@ -0,0 +1,155 @@
+// Copyright 2024 The Cockroach Authors.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package tests
+
+import (
+	"context"
+	"fmt"
+	"math/rand"
+	"strings"
+	"time"
+
+	"github.com/cockroachdb/cockroach/pkg/cmd/roachtest/cluster"
+	"github.com/cockroachdb/cockroach/pkg/cmd/roachtest/registry"
+	"github.com/cockroachdb/cockroach/pkg/cmd/roachtest/roachtestutil/mixedversion"
+	"github.com/cockroachdb/cockroach/pkg/cmd/roachtest/spec"
+	"github.com/cockroachdb/cockroach/pkg/cmd/roachtest/test"
+	"github.com/cockroachdb/cockroach/pkg/roachprod/logger"
+	"golang.org/x/exp/maps"
+)
+
+func registerMultiRegionMixedVersion(r registry.Registry) {
+	regionToZones := map[string][]string{
+		"us-east1":        {"us-east1-b"},
+		"us-west1":        {"us-west1-b"},
+		"europe-west2":    {"europe-west2-b"},
+		"europe-central2": {"europe-central2-b"},
+	}
+
+	regions := maps.Keys(regionToZones)
+	var zones []string
+	for _, zs := range maps.Values(regionToZones) {
+		zones = append(zones, zs...)
+	}
+
+	const (
+		nodesPerRegion                  = 20
+		backgroundWarehousesPerRegion   = 30
+		mixedVersionWarehousesPerRegion = 50
+	)
+
+	r.Add(registry.TestSpec{
+		Name:    "multi-region/mixed-version",
+		Timeout: 12 * time.Hour,
+		Owner:   registry.OwnerTestEng,
+		Cluster: r.MakeClusterSpec(
+			len(regions)*nodesPerRegion+1, // add one workload node
+			spec.Geo(),
+			spec.GCEZones(strings.Join(zones, ",")),
+		),
+		EncryptionSupport: registry.EncryptionMetamorphic,
+		CompatibleClouds:  registry.OnlyGCE,
+		Suites:            registry.Suites(registry.Weekly),
+		Run: func(ctx context.Context, t test.Test, c cluster.Cluster) {
+			partitionConfig := fmt.Sprintf(
+				"--regions=%s --partitions=%d",
+				strings.Join(regions, ","), len(regions),
+			)
+
+			roachNodes := c.Range(1, c.Spec().NodeCount-1)
+			mvt := mixedversion.NewTest(
+				ctx, t, t.L(), c, roachNodes,
+				// Our current fixtures only work for 4-node clusters.
+				mixedversion.NeverUseFixtures,
+				// Allow migrations to run for a longer period of time due to
+				// added latency and cluster size.
+				mixedversion.UpgradeTimeout(30*time.Minute),
+				// There are known issues upgrading from older patch releases
+				// in MR clusters (e.g., #113908), so use the latest patch
+				// releases to avoid flakes.
+				mixedversion.AlwaysUseLatestPredecessors,
+			)
+
+			backgroundTPCCOpts := tpccOptions{
```

Comment:
It might be worth mentioning explicitly that the background workload has no expiration (i.e., `Duration` is elided). It may not be immediately obvious given all the other config. options.

Reply:
Good point, added a comment about this.